crossbeam-queue/src/array_queue.rs

```diff
@@ -216,6 +216,84 @@ impl<T> ArrayQueue<T> {
         }
     }
 
+    /// Attempts to push an element into the queue without deadlocking.
+    ///
+    /// If the queue is full *or* if another operation would cause a potentially endless retry loop
+    /// (if the other operation is stalled) the element is returned back as an error.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_queue::{ArrayQueue, PushError};
+    ///
+    /// let q = ArrayQueue::new(1);
+    ///
+    /// assert_eq!(q.push(10), Ok(()));
+    /// assert_eq!(q.push(20), Err(PushError(20)));
+    /// ```
+    pub fn push_lockless(&self, value: T) -> Result<(), PushError<T>> {
```

Comment:
The code is very much similar with `push()`.  Do you think can we somehow merge the code of the two functions?

Reply:
Yes, I did think this and tried to merge them. Unfortunately it made the tests hang... So I will have another go at that when I have time. Before it is merged that should be done though.